In [ ]:
!pip install opendatasets --quiet

In [ ]:
import opendatasets as od

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
url = 'https://www.kaggle.com/datasets/tmdb/tmdb-movie-metadata?select=tmdb_5000_movies.csv'

In [ ]:
od.download(url)
#user_name = manishkumar7432698
#api = f60d7dd73290d421e95c6d0342be26d3

Please provide your Kaggle credentials to download this dataset. Learn more: http://bit.ly/kaggle-creds
Your Kaggle username: manishkumar7432698
Your Kaggle Key: ··········


100%|██████████| 8.89M/8.89M [00:00<00:00, 119MB/s]

In [ ]:
movie_df = pd.read_csv('/content/tmdb-movie-metadata/tmdb_5000_movies.csv')
credits_df = pd.read_csv('/content/tmdb-movie-metadata/tmdb_5000_credits.csv')

In [ ]:
movie_df.sample(5)

budget                                             genres  \
4509     550000  [{"id": 18, "name": "Drama"}, {"id": 53, "name...   
1135   42000000  [{"id": 80, "name": "Crime"}, {"id": 18, "name...   
253   110000000  [{"id": 28, "name": "Action"}, {"id": 10752, "...   
3881    3200000  [{"id": 18, "name": "Drama"}, {"id": 10749, "n...   
3266    8000000                     [{"id": 35, "name": "Comedy"}]   

                                               homepage     id  \
4509                                                NaN  75986   
1135                                                NaN   1830   
253                         http://www.300themovie.com/  53182   
3881                 http://focusfeatures.com/beginners  55347   
3266  http://disney.go.com/disneypictures/prom/index...  51588   

                                               keywords original_language  \
4509  [{"id": 1523, "name": "obsession"}, {"id": 213...                en   
1135  [{"id": 2106, "name": "cold war"}, {"id": 6110...                en   
253   [{"id": 18712, "name": "based on graphic novel...                en   
3881  [{"id": 237, "name": "gay"}, {"id": 936, "name...                en   
3266  [{"id": 641, "name": "single parent"}, {"id": ...                en   

              original_title  \
4509            Love Letters   
1135             Lord of War   
253   300: Rise of an Empire   
3881               Beginners   
3266                    Prom   

                                               overview  popularity  \
4509  A story of love and obsession. A young radio p...    0.001586   
1135  Yuri Orlov is a globetrotting arms dealer and,...   34.828241   
253   Based on Frank Miller's latest graphic novel X...   71.510596   
3881  A young man is rocked by two announcements fro...   11.933449   
3266  At “Prom,” every couple has a story and no two...   11.837067   

                                   production_companies  \
4509  [{"name": "New World Pictures", "id": 1950}, {...   
1135  [{"name": "Saturn Films", "id": 831}, {"name":...   
253   [{"name": "Legendary Pictures", "id": 923}, {"...   
3881  [{"name": "Parts and Labor", "id": 6531}, {"na...   
3266  [{"name": "Walt Disney Pictures", "id": 2}, {"...   

                                   production_countries release_date  \
4509  [{"iso_3166_1": "US", "name": "United States o...   1983-04-01   
1135  [{"iso_3166_1": "FR", "name": "France"}, {"iso...   2005-09-16   
253   [{"iso_3166_1": "US", "name": "United States o...   2014-03-05   
3881  [{"iso_3166_1": "US", "name": "United States o...   2010-09-11   
3266  [{"iso_3166_1": "US", "name": "United States o...   2011-04-29   

        revenue  runtime                                   spoken_languages  \
4509          0     98.0           [{"iso_639_1": "en", "name": "English"}]   
1135   24127895    122.0  [{"iso_639_1": "ar", "name": "\u0627\u0644\u06...   
253   337580051    102.0           [{"iso_639_1": "en", "name": "English"}]   
3881    5332926    105.0  [{"iso_639_1": "en", "name": "English"}, {"iso...   
3266   10130000    104.0           [{"iso_639_1": "en", "name": "English"}]   

        status                                            tagline  \
4509  Released        Sometimes It's Right To Do the Wrong Thing.   
1135  Released             Where there's a will, there's a weapon   
253   Released                                  Seize your glory!   
3881  Released                      This is what love feels like.   
3266  Released  There are hundreds of nights in high school, b...   

                       title  vote_average  vote_count  
4509            Love Letters           4.6           4  
1135             Lord of War           7.0        1335  
253   300: Rise of an Empire           6.1        2397  
3881               Beginners           6.8         347  
3266                    Prom           5.7         134

The data is cluttered. It has lot of information but not organized well and it make difficult for us to analyze it. If we directly provide this data to the model, model will get confused in giving weights to attributes.Hence, We need preprocessing for extracting important and useful information.

In [ ]:
credits_df.sample(3)

movie_id                    title  \
3981     66942      Griff the Invisible   
2753       492     Being John Malkovich   
2042     19366  Josie and the Pussycats   

                                                   cast  \
3981  [{"cast_id": 3, "character": "Griff", "credit_...   
2753  [{"cast_id": 5, "character": "Craig Schwartz",...   
2042  [{"cast_id": 8, "character": "Josie McCoy", "c...   

                                                   crew  
3981  [{"credit_id": "52fe4751c3a368484e0c0541", "de...  
2753  [{"credit_id": "52fe4249c3a36847f80128f7", "de...  
2042  [{"credit_id": "5645da789251410a3f008cb3", "de...

In [ ]:
credits_df['crew'][0]

'[{"credit_id": "52fe48009251416c750aca23", "department": "Editing", "gender": 0, "id": 1721, "job": "Editor", "name": "Stephen E. Rivkin"}, {"credit_id": "539c47ecc3a36810e3001f87", "department": "Art", "gender": 2, "id": 496, "job": "Production Design", "name": "Rick Carter"}, {"credit_id": "54491c89c3a3680fb4001cf7", "department": "Sound", "gender": 0, "id": 900, "job": "Sound Designer", "name": "Christopher Boyes"}, {"credit_id": "54491cb70e0a267480001bd0", "department": "Sound", "gender": 0, "id": 900, "job": "Supervising Sound Editor", "name": "Christopher Boyes"}, {"credit_id": "539c4a4cc3a36810c9002101", "department": "Production", "gender": 1, "id": 1262, "job": "Casting", "name": "Mali Finn"}, {"credit_id": "5544ee3b925141499f0008fc", "department": "Sound", "gender": 2, "id": 1729, "job": "Original Music Composer", "name": "James Horner"}, {"credit_id": "52fe48009251416c750ac9c3", "department": "Directing", "gender": 2, "id": 2710, "job": "Director", "name": "James Cameron"},

Our metadata dataset is also cluttered and diffiult to analyze.
Preprocesssing is required.

The first thing is we should merge the two datasets.

In [ ]:
movie = movie_df.merge(credits_df, on = 'title')

In [ ]:
movie.columns

Index(['budget', 'genres', 'homepage', 'id', 'keywords', 'original_language',
       'original_title', 'overview', 'popularity', 'production_companies',
       'production_countries', 'release_date', 'revenue', 'runtime',
       'spoken_languages', 'status', 'tagline', 'title', 'vote_average',
       'vote_count', 'movie_id', 'cast', 'crew'],
      dtype='object')

In [ ]:
movie.shape

(4809, 23)

Our data has lot of information and attributes and in some of the attributes also we have lot of data. Therefore,we need some relaxation from too much information. We will filter out some data according to our requirement means the data which help us in recommendation system.

Like no-one search a movie by its camera man name or any secondary crew. Hence, we will remove this data from our data set.

So, here is the list which is required for recommending a movie to a user which is based on user searches and personal experience.


1.   Genre
2.   id ( not for recommendation but for setting up a website it is important
3.   keywords
4.   title
5.   overview
6.   production_company( will use it later currently it will create complexity)
7.   release_date(currently not using but an important factor)
8.   cast
9.   crew(due to director)

In [ ]:
movies = movie[['movie_id','title','genres','keywords','overview','cast','crew']]

To create a recommendation system, I will create a new dataframe in which we will have only 3 attributes
1.   movie_id
2.   title
3.   Tag

In tag, we will have combination of all other columns apart from movie_id and title from movies dataset. But for this purpose, first I need to clean different columns and extract only required keywords from those attributes.



## **Data Preprocessing**

Let's do the above mentioned but to perform that, we need to clean and reduce cluterness of data.

### **Handling Null values**

In [ ]:
movies.isna().sum()

movie_id    0
title       0
genres      0
keywords    0
overview    3
cast        0
crew        0
dtype: int64

In [ ]:
movies = movies.dropna()

In [ ]:
movies.isna().sum()

movie_id    0
title       0
genres      0
keywords    0
overview    0
cast        0
crew        0
dtype: int64

### **Checking** **duplicates**

In [ ]:
movies.duplicated().sum()

0

### **Formatting of attributes**

In [ ]:
movies.head(1)

movie_id   title                                             genres  \
0     19995  Avatar  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                            keywords  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...

In [ ]:
import ast

ast module converts string to list.

In [ ]:
def extract(obj):
  l = []
  for i in ast.literal_eval(obj):
    l.append(i['name'])
  return l

In [ ]:
movies['genres'] = movies.genres.apply(extract)

In [ ]:
movies['genres']

0       [Action, Adventure, Fantasy, Science Fiction]
1                        [Adventure, Fantasy, Action]
2                          [Action, Adventure, Crime]
3                    [Action, Crime, Drama, Thriller]
4                [Action, Adventure, Science Fiction]
                            ...                      
4804                        [Action, Crime, Thriller]
4805                                [Comedy, Romance]
4806               [Comedy, Drama, Romance, TV Movie]
4807                                               []
4808                                    [Documentary]
Name: genres, Length: 4806, dtype: object

In [ ]:
movies['keywords'] = movies.keywords.apply(extract)

In [ ]:
def extract_actor(obj):
  l = []
  count = 0
  for i in ast.literal_eval(obj):
    if count < 3:
      l.append(i['name'])
      count+=1
    else:
       return l

In [ ]:
movies.cast = movies.cast.apply(extract_actor)

In [ ]:
movies.crew[0]

'[{"credit_id": "52fe48009251416c750aca23", "department": "Editing", "gender": 0, "id": 1721, "job": "Editor", "name": "Stephen E. Rivkin"}, {"credit_id": "539c47ecc3a36810e3001f87", "department": "Art", "gender": 2, "id": 496, "job": "Production Design", "name": "Rick Carter"}, {"credit_id": "54491c89c3a3680fb4001cf7", "department": "Sound", "gender": 0, "id": 900, "job": "Sound Designer", "name": "Christopher Boyes"}, {"credit_id": "54491cb70e0a267480001bd0", "department": "Sound", "gender": 0, "id": 900, "job": "Supervising Sound Editor", "name": "Christopher Boyes"}, {"credit_id": "539c4a4cc3a36810c9002101", "department": "Production", "gender": 1, "id": 1262, "job": "Casting", "name": "Mali Finn"}, {"credit_id": "5544ee3b925141499f0008fc", "department": "Sound", "gender": 2, "id": 1729, "job": "Original Music Composer", "name": "James Horner"}, {"credit_id": "52fe48009251416c750ac9c3", "department": "Directing", "gender": 2, "id": 2710, "job": "Director", "name": "James Cameron"},

I have to only extract the name of director.

In [ ]:
def extract_director(obj):
  l = []
  for i in ast.literal_eval(obj):
    if i['job'] == 'Director':
      l.append(i['name'])
      break
  return l

In [ ]:
movies['crew'] = movies['crew'].apply(extract_director)

In [ ]:
movies.overview = movies.overview.apply(lambda x:x.split())

In [ ]:
movies.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   
3               [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, Science Fiction]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   
2  [spy, based on novel, secret agent, sequel, mi...   
3  [dc comics, crime fighter, terrorist, secret i...   
4  [based on novel, mars, medallion, space travel...   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   
2  [A, cryptic, message, from, Bond’s, past, send...   
3  [Following, the, death, of, District, Attorney...   
4  [John, Carter, is, a, war-weary,, former, mili...   

                                               cast                 crew  
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]      [James Cameron]  
1     [Johnny Depp, Orlando Bloom, Keira Knightley]     [Gore Verbinski]  
2      [Daniel Craig, Christoph Waltz, Léa Seydoux]         [Sam Mendes]  
3      [Christian Bale, Michael Caine, Gary Oldman]  [Christopher Nolan]  
4    [Taylor Kitsch, Lynn Collins, Samantha Morton]     [Andrew Stanton]

## **Removing White space**
One more thing, I have to do here is to remove white spaces between the elements of any list in our dataset.

The reason why I am doing this is that to remove confusion to our model. Let's take the very first actor name which is `sam worthington` and the name of director of movie spectre is is `sam mendes`. If someone puts sam worthington in our recommender system then our system will search firstly for sam and also recommend a movie which is not of our consumers use. So to cure it, we will remove white spaces.

In [ ]:
def remove_space(columns):
  for column in columns:
    movies[column] = movies[column].apply(lambda x: [i.replace(" ",'') for i in x])

In [ ]:
columns = ['genres','keywords','cast','crew']
remove_space(columns)

TypeError: ignored

In [ ]:
movies.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                         genres  \
0  [Action, Adventure, Fantasy, ScienceFiction]   
1                  [Adventure, Fantasy, Action]   
2                    [Action, Adventure, Crime]   
3              [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, ScienceFiction]   

                                            keywords  \
0  [cultureclash, future, spacewar, spacecolony, ...   
1  [ocean, drugabuse, exoticisland, eastindiatrad...   
2  [spy, basedonnovel, secretagent, sequel, mi6, ...   
3  [dccomics, crimefighter, terrorist, secretiden...   
4  [basedonnovel, mars, medallion, spacetravel, p...   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   
2  [A, cryptic, message, from, Bond’s, past, send...   
3  [Following, the, death, of, District, Attorney...   
4  [John, Carter, is, a, war-weary,, former, mili...   

                                               cast                 crew  
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]      [James Cameron]  
1     [Johnny Depp, Orlando Bloom, Keira Knightley]     [Gore Verbinski]  
2      [Daniel Craig, Christoph Waltz, Léa Seydoux]         [Sam Mendes]  
3      [Christian Bale, Michael Caine, Gary Oldman]  [Christopher Nolan]  
4    [Taylor Kitsch, Lynn Collins, Samantha Morton]     [Andrew Stanton]

In [ ]:
movies.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                         genres  \
0  [Action, Adventure, Fantasy, ScienceFiction]   
1                  [Adventure, Fantasy, Action]   
2                    [Action, Adventure, Crime]   
3              [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, ScienceFiction]   

                                            keywords  \
0  [cultureclash, future, spacewar, spacecolony, ...   
1  [ocean, drugabuse, exoticisland, eastindiatrad...   
2  [spy, basedonnovel, secretagent, sequel, mi6, ...   
3  [dccomics, crimefighter, terrorist, secretiden...   
4  [basedonnovel, mars, medallion, spacetravel, p...   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   
2  [A, cryptic, message, from, Bond’s, past, send...   
3  [Following, the, death, of, District, Attorney...   
4  [John, Carter, is, a, war-weary,, former, mili...   

                                            cast                crew  
0  [SamWorthington, ZoeSaldana, SigourneyWeaver]      [JamesCameron]  
1     [JohnnyDepp, OrlandoBloom, KeiraKnightley]     [GoreVerbinski]  
2      [DanielCraig, ChristophWaltz, LéaSeydoux]         [SamMendes]  
3      [ChristianBale, MichaelCaine, GaryOldman]  [ChristopherNolan]  
4    [TaylorKitsch, LynnCollins, SamanthaMorton]     [AndrewStanton]

## **Creation of Tag column**
Last thing to do for processing is to create a tag column which is combination genres, keywords, overview, cast, crew. The tag column will be in small cap and in form of string.

In [ ]:
movies['tags'] = movies['overview'] + movies['genres'] + movies['keywords'] + movies['cast'] + movies['crew']

In [ ]:
new_df = movies[['movie_id','title','tags']]

In [ ]:
new_df['tags'] = new_df['tags'].apply(lambda x: " ".join(x))

TypeError: ignored

In [ ]:
new_df['tags'] = new_df['tags'].apply(lambda x: x.lower())

AttributeError: ignored

In [ ]:
new_df.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                tags  
0  in the 22nd century, a paraplegic marine is di...  
1  captain barbossa, long believed to be dead, ha...  
2  a cryptic message from bond’s past sends him o...  
3  following the death of district attorney harve...  
4  john carter is a war-weary, former military ca...

Here, we need to do stemming to avoid repetition of words in different forms in most common words. It will improve the accuracy very much.

In [ ]:
import nltk

In [ ]:
from nltk.stem.porter import PorterStemmer
ps = PorterStemmer()

In [ ]:
def stem(text):
  y = []

  for i in text.split():
    y.append(ps.stem(i))
  return ' '.join(y)

In [ ]:
new_df['tags'] = new_df.tags.apply(stem)

<ipython-input-49-33874492b6a7>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df['tags'] = new_df.tags.apply(stem)


In [ ]:
new_df

movie_id                                     title  \
0        19995                                    Avatar   
1          285  Pirates of the Caribbean: At World's End   
2       206647                                   Spectre   
3        49026                     The Dark Knight Rises   
4        49529                               John Carter   
...        ...                                       ...   
4804      9367                               El Mariachi   
4805     72766                                 Newlyweds   
4806    231617                 Signed, Sealed, Delivered   
4807    126186                          Shanghai Calling   
4808     25975                         My Date with Drew   

                                                   tags  
0     in the 22nd century, a parapleg marin is dispa...  
1     captain barbossa, long believ to be dead, ha c...  
2     a cryptic messag from bond’ past send him on a...  
3     follow the death of district attorney harvey d...  
4     john carter is a war-weary, former militari ca...  
...                                                 ...  
4804  el mariachi just want to play hi guitar and ca...  
4805  a newlyw couple' honeymoon is upend by the arr...  
4806  "signed, sealed, delivered" introduc a dedic q...  
4807  when ambiti new york attorney sam is sent to s...  
4808  ever sinc the second grade when he first saw h...  

[4806 rows x 3 columns]

## **Modelling**
How our model will work -

1. Create a bag of words using all available text.
2. Extract the most common words to be used as the vectorizer.
3. Pass the tags column to the vectorizer.
4. Convert the defined number of text (5000) in the vectorizer into column headings, with the movie names as rows.
5. Analyze the movies and assign a value of 1 to each column heading that is present in the movie tag text.
6. Plot a graph based on the analysis.
7. If someone searches for a word, suggest the 5 or 10 closest points on the graph to the searched word.

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer

In [ ]:
cv = CountVectorizer(max_features = 5000, stop_words = 'english')

In [ ]:
vectors = cv.fit_transform(new_df['tags']).toarray()

In [ ]:
vectors

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

This is our points which we will plot on the graph, every row of this matrix represent the matched words in the first movie and the most common words in our bag of words.

In [ ]:
# cv.get_feature_names()

Improvement on 1st run - Now, We will apply stemming to remove or merge similar words. Like in our data words such as actor and actors are treated differently but they are same. So to remove this we do stemming.So, we have to go back for now to remove these words from the root column which was nltk.

In [ ]:
vectors.shape

(4806, 5000)

In this shape, 4806 is representing the number of movies available to the machine and 5000 is representing the most common 5000 words from the bag of words
.

## **Distance Calculation**
After vectorization is calculating the distance between the movies on this 5000*5000 dimensional graph. It will help us to suggest the similar movies. Here, we will use cosine distane not the eucleadian distance because the more the dimension, the bad the eucleadean distance for the method.

In [ ]:
from sklearn.metrics.pairwise import cosine_similarity

In [ ]:
similarity = cosine_similarity(vectors)

This array is representing the distance(dissimilarity) of movies with each other. Its shape is 4806X4806 which is verifying the statement.

In [ ]:
similarity

array([[1.        , 0.08346223, 0.0860309 , ..., 0.04499213, 0.        ,
        0.        ],
       [0.08346223, 1.        , 0.06063391, ..., 0.02378257, 0.        ,
        0.02615329],
       [0.0860309 , 0.06063391, 1.        , ..., 0.02451452, 0.        ,
        0.        ],
       ...,
       [0.04499213, 0.02378257, 0.02451452, ..., 1.        , 0.03962144,
        0.04229549],
       [0.        , 0.        , 0.        , ..., 0.03962144, 1.        ,
        0.08714204],
       [0.        , 0.02615329, 0.        , ..., 0.04229549, 0.08714204,
        1.        ]])

The similarity of a movie with itself is obviously one and that is represented by the diagonals.

## **Recommendation Function**

In [ ]:
def recommend(movie):
  movie_index = new_df[new_df['title'] == movie].index[0]
  similar = sorted(list(enumerate(similarity[movie_index])),reverse = True,key = lambda x: x[1])[1:6]
  for i in similar:
    print(new_df.iloc[i[0]].title)

In [ ]:
recommend('Avatar')

Aliens vs Predator: Requiem
Aliens
Falcon Rising
Independence Day
Titan A.E.


In [ ]:
import joblib
import gzip

In [ ]:
# with gzip.open('similarity.pkl', 'wb') as f:
#     pickle.dump(similarity, f)

In [ ]:
joblib.dump(new_df,open('movies.joblib','wb'))

In [ ]:
joblib.dump(similarity,open('similarity.pkl','wb'))

In [ ]:
import bz2

In [ ]:
with bz2.open('similarity.pkl', 'wb') as f:
    pickle.dump(similarity, f)

In [ ]:
import joblib

In [ ]:
joblib.dump(similarity, 'similarity.pkl', compress = 8)

['similarity.pkl']

In [ ]:
joblib.dump(similarity, 'similarity5.pkl', compress = 9)

['similarity5.pkl']